ambry-cloud/src/main/java/com.github.ambry.cloud/CloudMessageReadSet.java

```diff
@@ -0,0 +1,174 @@
+    validateIndex(index);
+    long written = 0;
+    BlobReadInfo blobReadInfo = blobReadInfoList.get(index);
+
+    try {
+      if (blobReadInfo.isPrefetched()) {
+        ByteBuffer outputBuffer = blobReadInfo.getPrefetchedBuffer();
+        outputBuffer.flip();
+        written = channel.write(outputBuffer);
+      } else {
+        blobReadInfo.downloadBlob(cloudDestination, Channels.newOutputStream(channel));
+        written = blobReadInfo.getBlobSize();
+      }
+    } catch (CloudStorageException ex) {
+      throw new IOException("Download of cloud blob " + blobReadInfoList.get(index).getBlobId().getID() + " failed");
+    }
+    logger.trace("Downloaded {} bytes to the write channel from the cloud blob : {}", written,
+        blobReadInfoList.get(index).getBlobId().getID());
+    return written;
+  }
+
+  @Override
+  public int count() {
+    return blobReadInfoList.size();
+  }
+
+  @Override
+  public long sizeInBytes(int index) {
+    validateIndex(index);
+    return blobReadInfoList.get(index).getBlobSize();
+  }
+
+  @Override
+  public StoreKey getKeyAt(int index) {
+    validateIndex(index);
+    return blobReadInfoList.get(index).getBlobId();
+  }
+
+  @Override
+  public void doPrefetch(int index, long relativeOffset, long size) throws IOException {
+    try {
+      blobReadInfoList.get(index).prefetchBlob(cloudDestination);
+    } catch (CloudStorageException ex) {
+      throw new IOException("Prefetch of cloud blob " + blobReadInfoList.get(index).getBlobId().getID() + " failed");
+    }
+  }
+
+  /**
+   * Validate that the index is withing the bounds of {@code blobReadInfoList}
+   * @param index The index to be verified
+   * @throws {@code IndexOutOfBoundsException} if the index is out of bounds
+   */
+  private void validateIndex(int index) {
+    if (index >= blobReadInfoList.size()) {
+      throw new IndexOutOfBoundsException("index [" + index + "] out of the messageset");
+    }
+  }
+
+  /**
+   * A class to maintain download information about each blob
+   */
+  static class BlobReadInfo {
+    private final CloudBlobMetadata blobMetadata;
+    private final BlobId blobId;
+    private ByteBuffer prefetchedBuffer;
+    private boolean isPrefetched;
+
+    public BlobReadInfo(CloudBlobMetadata blobMetadata, BlobId blobId) {
+      this.blobMetadata = blobMetadata;
+      this.blobId = blobId;
+      this.isPrefetched = false;
+    }
+
+    /**
+     * Prefetch the {@code blob} from {@code CloudDestination} and put it in {@code prefetchedBuffer}
+     * @param cloudDestination {@code CloudDestination} implementation representing the cloud from which download will happen.
+     * @throws CloudStorageException if blob cloud not be downloaded
+     */
+    public void prefetchBlob(CloudDestination cloudDestination) throws CloudStorageException {
+      prefetchedBuffer = ByteBuffer.allocate((int) blobMetadata.getSize());
```

Comment:
Minor: add comment explaining that the casting is okay because the blobs are chunked in the cloud, but that if we ever decide to upload entire blobs (and let Azure do the chunking for us), then the prefetching logic will have to change.

Reply:
fixed.